x-pack/plugins/security_solution/docs/prebuilt_rules_customization_rfc.md

```diff
@@ -0,0 +1,1357 @@
+
+
+_Source for `getExportAll`: [x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/export/get_export_all.ts](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/logic/export/get_export_all.ts)_
+
+### Importing rules
+
+The user will now be able to import both custom and prebuilt rules (including any customizations that might have been performed on its modifiable fields).
+
+We want to handle the case of importing both migrated and non-migrated rules via an `ndjson` file.
+
+
+If a user imports a prebuilt rule, Kibana should continue to track the rule asset from the `security_detection_engine` package if the `rule_id` matches. This means that a user will be able to update a previously exported prebuilt rule, visualize any diffs on any customized fields, and continue to receive updates on the rule.
+
+To allow for importing of Elastic prebuilt rules, we will **not rely** in the `rule_source_type` or `external_source` fields (which are not part of the import endpoint parameters), but we will rather **calculate them dynamically based on the `rule_id` and `version` request parameters**.
+
+The logic to importing a rule is as follows:
+
+First, read the import endpoint request parameters `rule_id` and `version`. These two are the two required parameters in the endpoint, while all other rule fields are optional.
+
+Secondly, check that the `security_detection_engine` Fleet package is installed and install it if it isn't. We will need the package to be installed to check if a rule that is being imported is an Elastic prebuilt rule.
+
+Then, using the `rule_id` and `version`, attempt to fetch the corresponding `security-rule` asset from ES. 
+
+**If a matching `rule_id` and `version` is found**, that means that the rule is an Elastic prebuilt rule, and we should therefore dynamically calculate the rule's `external_source` and `rule_source_type` fields.
+
+- `rule_source_type`: should always be `'external'` if rule is Elastic prebuilt.
+- `external_source`:
+  - `source_updated_at`: can be retrieved from the corresponding `security-rule` asset.
+  - `is_customized`: should be calculated based on the `security-rule` asset's field and the field rules that were part of the import request parameters. If any of them are different, i.e. have diverged from the base version, `is_customized` should be true.
+
+Finally, using the import payload, plus the rule's `security-rule` asset fields and the calculated `rule_source_type` and `external_source` fields, create the rule, or update it if already exists in Kibana.
+
+**If a matching `rule_id` is found, but the `version` is not found**, reject the import for that rule. Without a version, we cannot dynamically calculate `external_source`.
```

Comment:
This is a case which we didn't discuss in the last Product meeting: what if a user sends as import parameters a `rule_id` that matches a `security-rule` asset, but the `version` doesn't match any historical (or the last) version?

Given that we've decided to "pay the cost" of making sure that the package is installed, I have tentatively decided here that the way to handle this edge case is to reject the import, clearly specifying why. Otherwise, we will fall in the pitfalls that @marshallmain explained [in this comment.](https://github.com/elastic/kibana/pull/171856#discussion_r1491794090)

**But** I'm not fully convinced this is the best UX, as a user won't be able to know which rule versions exist in Kibana, so not sure how they would attempt to "fix" this issue if they fell into this situation.

cc @banderror

Reply:
@jpdjere Commented on that in https://github.com/elastic/kibana/pull/171856#discussion_r1524431230